### PR TITLE
Add Zihintntl spec

### PR DESCRIPTION
**Pull request by aswaterman**
Non-temporal locality hints.  See text.

### CODE REVIEW

**Comment by allenjbaum** [comment]
Nice. but...corner cases.
Is the intent that if the load/store is a vector load/store, and traps in
the middle, that it affects only (and all of) the first part? (so, multiple
cache lines)
IF it is an unaligned store with HW support, yet still traps, and the
implementation is such that it will store the first part before trapping,
    then does first part of the access treated as non-temporal, but not the
second? (a restricted version of the above)
Both of those seem kind of obvious.

IS there any sidechannel leakage that occurs because of that behavior, i.e.
timing an interrupt causing something to be cached vs. not cached,
with the resulting timing differences being seen later?

I note that you refer to zicbop, which is not ratified yet, and referred to
nowhere in (this) spec version.
 Assuming there will eventually be different flavors of prefetch (read
intent, write intent, instruction)
- do they get treated any differently from each other beyond what is
described here?

File: src/zihintntl.tex [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+If the C extension is provided, compressed variants of these HINTs are also
+provided:
+C.NTL.P1 is encoded as \mbox{C.ADD {\em x0, x2}};
+C.NTL.PALL is encoded as \mbox{C.ADD {\em x0, x3}};
+C.NTL.S1 is encoded as \mbox{C.ADD {\em x0, x4}};
+and C.NTL.ALL is encoded as \mbox{C.ADD {\em x0, x5}}.
+
+The NTL instructions affect all memory-access instructions, unless otherwise
+specified.
```
Comment (dkruckemyer-ventana) [inline src/zihintntl.tex:142]: They must not affect management CBOs (Zicbom). Zero and prefetch CBOs (as you state below) can be affected.

File: src/zihintntl.tex [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+\begin{commentary}
+As of this writing, there are no exclusions to this rule, and so
+the NTL instructions affect all memory-access instructions
+defined in the base ISAs and the A, F, D, Q, C, and V standard extensions,
+as well as those defined within the hypervisor extension in Volume II.
+
+We also expect that the NTL instructions will affect some cache-maintenance
```
Comment (dkruckemyer-ventana) [inline src/zihintntl.tex:150]: "management" is the preferred term... :)

File: src/zihintntl.tex [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+operations.
+For example, NTL.PALL followed by a cache-line-zero instruction might indicate
```
Comment (dkruckemyer-ventana) [inline src/zihintntl.tex:152]: CBO.ZERO instead of "cache-line-zero" to be precise?

**Comment by pdonahue-ventana** [comment]
> I note that you refer to zicbop, which is not ratified yet

https://wiki.riscv.org/display/TECH/Recently+Ratified+Extensions says that Zicbop was ratified in November.

**Comment by allenjbaum** [comment]
Huh -I based that comment on searching for it in the latest 20191213 unpriv
spec and the 20211203 priv spec  but failed to check the Recently Ratified
Extensions page.
My apologies.

**Comment by aswaterman** [comment]
> Is the intent that if the load/store is a vector load/store, and traps in the middle, that it affects only (and all of) the first part? (so, multiple cache lines)

Because it's a HINT, this is up to the implementation as long as there isn't an architecturally visible effect, and so we're unburdened from having to specify this.  In practice, if an implementation follows our recommendations, the NTL will stop taking effect upon trapping.  Then, when execution resumes at the faulting load or store, the NTL won't be re-executed, and so won't apply to the remainder.

> IS there any sidechannel leakage that occurs because of that behavior, i.e. timing an interrupt causing something to be cached vs. not cached, with the resulting timing differences being seen later?

Yes, but this could've been observed more readily through other means (e.g., reading the timer before and after).

File: src/zihintntl.tex (hunk 1 above)
Comment (aswaterman) [reply src/zihintntl.tex:142]: Ack.

File: src/zihintntl.tex (hunk 2 above)
Comment (aswaterman) [reply src/zihintntl.tex:150]: Ack.

File: src/zihintntl.tex (hunk 3 above)
Comment (aswaterman) [reply src/zihintntl.tex:152]: Ack.

**Comment by allenjbaum** [comment]
OK, I reread it, and notice recommendations and discouragements, but no
real requirements or prohibitions.
An implementation with decoded Icache could even encode the hint there, so
re-execution or re-start could re-use the hint.
I'm not totally convinced with the side-channel argument, but because I
don't have that malware mindset, I can't justify that non-conviction.
Looks good.

**Pull request merged**